service/storage_service.cc

```diff
@@ -2457,6 +2454,50 @@ future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>&
     for (auto& x : loaded_peer_features) {
         slogger.info("peer={}, supported_features={}", x.first, x.second);
     }
+
+    if (!_raft_experimental_topology) {
+        slogger.info("Booting in legacy operations mode because experimental raft topology is not enabled");
+        _topology_change_kind_enabled = topology_change_kind::legacy;
+    } else if (utils::get_local_injector().is_enabled("force_gossip_based_join") && !_sys_ks.local().bootstrap_complete()) {
+        slogger.info("Booting in legacy topology operations mode because it was requested by an error injection");
+        _topology_change_kind_enabled = topology_change_kind::legacy;
+    } else if (_group0->client().in_recovery()) {
+        slogger.info("Raft recovery - starting in legacy topology operations mode");
+        _topology_change_kind_enabled = topology_change_kind::legacy;
+    } else if (_group0->joined_group0()) {
+        slogger.info("Restarting in raft mode");
+        _topology_change_kind_enabled = topology_change_kind::raft;
+    } else if (_sys_ks.local().bootstrap_complete()) {
+        // We already bootstrapped but we are not a part of group 0. This means that we are restarting after recovery.
+        slogger.info("Restarting after recovery - starting in legacy topology operations mode");
```

Comment:
Could be restarting with upgrade (5.4 with Raft disabled -> 6.0) not just recovery